### ISSUE TITLE
Extension causes high cpu load

### ISSUE THREAD

**Issue by ClimenteA** [sveltejs/language-tools#642]
- Issue Type: `Performance`
- Extension Name: `svelte-vscode`
- Extension Version: `102.3.0`
- OS Version: `Linux x64 5.4.0-52-generic`
- VSCode version: `1.50.1`

:warning: Make sure to **attach** this file from your *home*-directory:
:warning:`/tmp/svelte.svelte-vscode-unresponsive.cpuprofile.txt`

Find more details here: https://github.com/microsoft/vscode/wiki/Explain-extension-causes-high-cpu-load

**Comment by dummdidumm** [comment]
Please upload the cpuprfile file or we cannot investigate this.

**Comment by ClimenteA** [comment]
When I open vscode there is a message notification message at the bottom that says:
"Initializing JS/TS language features" (this keeps running all the time)

in what situation did this occur (for example after a file rename)
At one moment a dialog appeared and suggested to open an issue (noticed that my pc was pretty slow yesterday (intel i5) )

how long did you have the Svelte project open at that time already?
3-4 hours

Size of your project (nr of Svelte files + nr of JS/TS/other files)
Pretty small, about 10 files under 100 LOC (just finished a landing page)

Here is my package.json (maybe is something else idk)
```
{
  "name": "svelte-app",
  "version": "1.0.0",
  "scripts": {
    "watch:tailwind": "postcss public/tailwind.css -o public/index.css -w",
    "build:tailwind": "cross-env NODE_ENV=production postcss public/tailwind.css -o public/index.css",
    "build": "npm run build:tailwind && rollup -c",
    "start": "sirv public",
    "serve": "sirv public -p 80",
    "dev": "concurrently \"rollup -c -w\" \"npm run watch:tailwind\""
  },
  "devDependencies": {
    "@rollup/plugin-commonjs": "^14.0.0",
    "@rollup/plugin-node-resolve": "^8.0.0",
    "autoprefixer": "^9.8.6",
    "concurrently": "^5.3.0",
    "cross-env": "^7.0.2",
    "postcss-cli": "^8.1.0",
    "rollup": "^2.3.4",
    "rollup-plugin-livereload": "^2.0.0",
    "rollup-plugin-svelte": "^6.0.0",
    "rollup-plugin-terser": "^7.0.0",
    "svelte": "^3.0.0",
    "tailwindcss": "^1.9.6"
  },
  "dependencies": {
    "sirv-cli": "^1.0.0",
    "svelte-spa-router": "^3.0.5"
  }
}

```

**Comment by dummdidumm** [comment]
Closing because not reproducible. If you got anything new, just comment here or open another issue.

**Comment by rthomps7** [comment]
I'm seeing issues here. I have a workspace that has a few sapper projects in it.

If I watch the logs of my svelte plugin, randomly at some point it will show this:

```
Initialize new ts service at  
Trying to load configs for
```

Where it should normally have a filename at the end of each line.

It basically hangs up my machine. Sometimes my machine can get through it and sometimes it takes my machine down.

I often kill the process so I can keep moving but it's obviously a pain.

Curious if anyone has an idea what's up

**Comment by dummdidumm** [comment]
Without any details there's nothing we can do. Ideally you can provide a reproducible or the link to the repo. Else:
- what's the folder structure?
- how many projects/svelte files/js/TS files are we talking about?
- is there a jsconfig or tsconfig in these projects?

**Comment by rthomps7** [comment]
Yeah - understood. It's not something I can share since it's private. I want to take a look at creating a reproducible situation or look at solving myself, but figured I'd check if this made any sense here first.

Something like:

```
projects/
  jsconfig.json // tried adding this to tell it to ignore everything if it got to this level
  project1/
      jsconfig.json
      ... other files
  project2/
     tsconfig.json // one of them is typescript
     ... other files
  project3/
    jsconfig.json
```

I'm never usually working in all projects at once, but often two because they communicate, which is why they aren't just separate workspaces. 

Understandably, if the svelte plugin needs to watch 100 projects at some point it would break down. But it's very confusing to me what's triggering the initialization with no path. And I've found no pattern to when that gets fired off yet. It can even happen if I'm doing nothing and come back to vscode.

**Comment by dummdidumm** [comment]
The situation of "at _blank_" happens when the language service does not find a file inside the project. It traverses up to the project root and stops there. Since you said there's a config in each project something weird seems to happen.

**Comment by rthomps7** [comment]
That makes sense. Does it trigger only on saved files (that it then can't find) or does it trigger related to imported files or those found from scanning for imports/types/etc?

The `jconfig.json` at my root is ignoring all project folders, so I was hoping that would catch it and stop, but perhaps it fails cause the file that is being looked at is not inside it.

I'll keep digging and see if I can find which file is causing it and go from there. Thanks for the direction

**Comment by dummdidumm** [comment]
For every file that is loaded into the language service, it tries to find the related jsconfig/tsconfig (up to the workspace root). If it doesn't find one, it will fall back to the empty one. What's strange to me is that it should show "loading new at _blank_" only once, because then that service location is saved and others that also resolve to _blank_ should use that service, too.

The log (in which you find that "loading new at ...") also prints out a list of "SnapshotManager File Statistics" from time to time which might be interesting, too. Maybe it's unusually high which hints at possible problems, too.

**Comment by rthomps7** [comment]
Thanks @dummdidumm 

It does only happen once. But somethings I kill the process because it'll take down my machine. Then it'll try again later. 

I've never gotten good SnapshotManager stats because my machine can't make it through loading everything with no ignores. 

At any rate, I'll sort it out and then update this thread if I find something. Thanks for the direction

**Comment by rthomps7** [comment]
@dummdidumm 

This may be interesting to you - apologies if I'm just bothersome. But I have discovered which file it's trying to open:

```
~/.vscode-server/extensions/svelte.svelte-vscode-105.3.1/node_modules/svelte-language-server/node_modules/svelte2tsx/svelte-jsx.d.ts
```

Still debugging but I did find that interesting.

**Comment by dummdidumm** [comment]
This file is loaded on startup of a new service. The startup logic is here: https://github.com/sveltejs/language-tools/blob/master/packages/language-server/src/plugins/typescript/service.ts#L89
Basically it loads some ambient type definitions and the content of the tsconfig and then tells typescript to find all file names that match the tsconfig. This is likely the point at which it breaks for you because it loads too many files.
If you want to keep debugging I suggest to set up the extension locally, then you can set break points when attaching the debugger. More info on how to set it up here: https://github.com/sveltejs/language-tools#development

**Comment by guohaolay** [comment]
I can reproduce this too, simply open up a `.ts` file, followed by a `.svelte` file and you'll see the spinner go on forever.
